File: mycollab-web/src/main/java/com/esofthead/mycollab/vaadin/mvp/AbstractPresenter.java
package com.esofthead.mycollab.vaadin.mvp;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.esofthead.mycollab.common.domain.PermissionMap;
import com.esofthead.mycollab.core.utils.BeanUtility;
import com.esofthead.mycollab.security.PermissionChecker;
import com.esofthead.mycollab.vaadin.ui.MessageConstants;
import com.esofthead.mycollab.web.AppContext;
import com.vaadin.ui.ComponentContainer;

public abstract class AbstractPresenter<V extends View> implements Presenter<V> {

	private static final long serialVersionUID = 1L;
	private static Logger log = LoggerFactory
			.getLogger(AbstractPresenter.class);
	protected V view;

	public AbstractPresenter(Class<V> viewClass) {
		view = ViewManager.getView(viewClass);
	}

	@Override
	public V getView() {
		return view;
	}

	@Override
	public void go(ComponentContainer container, ScreenData<?> data) {
		go(container, data, true);
	}

	@Override
	public void go(ComponentContainer container, ScreenData<?> data,
			boolean isHistoryTrack) {
		log.debug("Go to view: " + view);
		if (isHistoryTrack) {
			ViewState state = new ViewState(container, this, data);
			if (log.isDebugEnabled()) {
				StringBuilder str = new StringBuilder();
				str.append("----").append("\n");
				str.append("Add ").append(state).append("\n");
				str.append("to history with container ").append(container)
						.append("\n");
				str.append("----");
				log.debug(str.toString());
			}
			HistoryViewManager.addHistory(state);
		} else {
			log.debug("Disable history track for " + this);
		}

		if (checkPermissionAccessIfAny()) {
			onGo(container, data);
		} else {
			MessageConstants.showMessagePermissionAlert();
		}

	}

	protected abstract void onGo(ComponentContainer container,
			ScreenData<?> data);

	private boolean checkPermissionAccessIfAny() {
		ViewPermission viewPermission = this.getClass().getAnnotation(
				ViewPermission.class);
		if (viewPermission != null) {
			String permissionId = viewPermission.permissionId();
			int impliedPermissionVal = viewPermission.impliedPermissionVal();

			if (AppContext.isAdmin()) {
				return true;
			} else {
				PermissionMap permissionMap = AppContext.getPermissionMap();
				if (permissionMap == null) {
					return false;
				} else {
					Integer value = permissionMap.get(permissionId);
					if (value == null) {
						return false;
					} else {
						return PermissionChecker.isImplied(value,
								impliedPermissionVal);
					}
				}
			}
		} else {
			return true;
		}
	}

	@Override
	public void handleChain(ComponentContainer container,
			PageActionChain pageActionChain) {
		log.debug("Handle chain: " + BeanUtility.printBeanObj(pageActionChain)
				+ "---" + this);
		ScreenData pageAction = pageActionChain.pop();
		go(container, pageAction);

		if (pageActionChain.hasNext()) {
			onHandleChain(container, pageActionChain);
		} else {
			onDefaultStopChain();
		}
	}

	protected void onDefaultStopChain() {

	}

	protected void onHandleChain(ComponentContainer container,
			PageActionChain pageActionChain) {
		throw new UnsupportedOperationException("You need override this method");
	}
}

File: mycollab-mobile/src/main/java/com/mycollab/mobile/mvp/AbstractPresenter.kt
package com.mycollab.mobile.mvp

import com.mycollab.core.ResourceNotFoundException
import com.mycollab.core.SecureAccessException
import com.mycollab.core.utils.ExceptionUtils.getExceptionType
import com.mycollab.security.PermissionChecker
import com.mycollab.vaadin.UserUIContext
import com.mycollab.vaadin.mvp.*
import com.mycollab.vaadin.ui.NotificationUtil
import com.vaadin.addon.touchkit.ui.NavigationManager
import com.vaadin.ui.HasComponents
import com.vaadin.ui.UI
import org.slf4j.LoggerFactory

/**
 * @param <V>
 * @author MyCollab Ltd.
 * @since 3.0
</V> */
abstract class AbstractPresenter<out V : PageView>(private val viewClass: Class<V>) : IPresenter<V> {
    private var _view: V? = null

    override fun getView(): V? {
        if (_view == null) {
            _view = ViewManager.getCacheComponent(viewClass)
            postInitView()
        }
        return _view
    }

    protected open fun postInitView() {}

    fun go(data: ScreenData<Any>) {
        val manager = UI.getCurrent().content as NavigationManager
        go(manager, data)
    }

    override fun go(container: HasComponents, data: ScreenData<Any>?): Boolean {
        getView()

        when {
            checkPermissionAccessIfAny() -> try {
                onGo(container, data)
            } catch (e: Throwable) {
                onErrorStopChain(e)
                return false
            }
            else -> NotificationUtil.showMessagePermissionAlert()
        }
        return true
    }

    protected abstract fun onGo(container: HasComponents, data: ScreenData<Any>?)

    private fun checkPermissionAccessIfAny(): Boolean {
        val viewPermission = this.javaClass.getAnnotation(ViewPermission::class.java)
        when {
            viewPermission != null -> return if (UserUIContext.isAdmin()) {
                true
            } else {
                val permissionId = viewPermission.permissionId
                val impliedPermissionVal = viewPermission.impliedPermissionVal
                val permissionMap = UserUIContext.getPermissionMap()
                when (permissionMap) {
                    null -> false
                    else -> {
                        val value = permissionMap[permissionId]
                        value != null && PermissionChecker.isImplied(value, impliedPermissionVal)
                    }
                }
            }
            else -> return true
        }
    }

    override fun handleChain(container: HasComponents, pageActionChain: PageActionChain) {
        val pageAction = pageActionChain.pop()
        val isSuccess = go(container, pageAction)

        if (pageActionChain.hasNext() && isSuccess) {
            onHandleChain(container, pageActionChain)
        }
    }

    protected open fun onErrorStopChain(throwable: Throwable) {
        when {
            getExceptionType(throwable, ResourceNotFoundException::class.java) != null -> NotificationUtil.showRecordNotExistNotification()
            getExceptionType(throwable, SecureAccessException::class.java) != null -> NotificationUtil.showMessagePermissionAlert()
            getExceptionType(throwable, PresenterNotFoundException::class.java) != null -> NotificationUtil.showFeatureNotPresentInSubscription()
            else -> LOG.error("Exception", throwable)
        }
    }

    protected open fun onHandleChain(container: HasComponents, pageActionChain: PageActionChain) {
        throw UnsupportedOperationException("You need override this method")
    }

    companion object {
        private val serialVersionUID = 1L
        private val LOG = LoggerFactory.getLogger(AbstractPresenter::class.java)
    }
}
